txpool/txpool.go

```diff
@@ -555,38 +576,76 @@ func (t *TxPool) Underpriced(tx *types.Transaction) bool {
 	return underpriced
 }
 
-func (t *TxPool) Discard(slots uint64, force bool) ([]*types.Transaction, bool) {
+func (t *TxPool) Discard(remaining uint64, force bool) ([]*types.Transaction, bool) {
 	dropped := make([]*types.Transaction, 0)
-	for t.remoteTxns.Length() > 0 && slots > 0 {
+	for t.remoteTxns.Length() > 0 && remaining > 0 {
 		tx := t.remoteTxns.Pop()
 		dropped = append(dropped, tx.tx)
 
-		txSlots := numSlots(tx.tx)
-		if slots >= txSlots {
-			slots -= txSlots
+		txSlots := slots(tx.tx)
+		if remaining >= txSlots {
+			remaining -= txSlots
 		} else {
-			slots = 0
+			remaining = 0
 		}
 	}
 
 	// Put back if couldn't make required space
-	if slots > 0 && !force {
+	if remaining > 0 && !force {
 		for _, tx := range dropped {
 			t.remoteTxns.Push(tx)
 		}
 		return nil, false
 	}
+
 	return dropped, true
 }
 
 // increaseSlots increases number of taken slots
 func (t *TxPool) increaseSlots(slots uint64) {
-	atomic.AddUint64(&t.slots, slots)
+	t.gauge.increase(slots)
+	// atomic.AddUint64(&t.slots, slots)
 }
 
 // increaseSlots decreases number of taken slots
 func (t *TxPool) decreaseSlots(slots uint64) {
-	atomic.AddUint64(&t.slots, ^(slots - 1))
+	t.gauge.decrease(slots)
+	// atomic.AddUint64(&t.slots, ^(slots - 1))
+}
+
+// Checks if the incoming tx would cause an overflow
+// and attempts to allocate space for it
+func (t *TxPool) gaugeCheck(tx *types.Transaction, isLocal bool) error {
+	t.gauge.lock()
+	defer t.gauge.unlock()
+
+	if t.gauge.current+slots(tx) <= t.gauge.limit {
+		// no overflow
+		return nil
+	}
+
+	if !isLocal && t.Underpriced(tx) {
```

Comment:
This doesn't seem related to the rest of the method code, this should be done in other place.

Reply:
Perhaps @Kourin1996 can provide more detail on the original idea behind that specific logic.

As it is right now, that part can be translated to: "In case of an overflow, no room will be made for remote transactions with lower gas fee than the transaction with the lowest fee currently present in the pool".  Resulting in a dropped transaction.

It's a decision-making move tied to the event of an overflow. I'm not entirely sure if it should be put elsewhere, but it definitely call for a discussion.

Reply:
I think we should leave it in this method as it is now, since it was all in `addImpl` anyways, we just moved it to a helper.
Moving it out as the txpool currently stands would just complicate spaghetti the logic even more.

When we finally do a refactor of the txpool, this will definitely get outsourced to a more appropriate place